### ISSUE TITLE
Error : failed to restore any of the available snapshots

### ISSUE THREAD

**Issue by pinaki124** [hashicorp/consul#6991]
#### Overview of the Issue

Running a 3 node Consul 'server' cluster. The cluster had been running and operating fine while the consul agent process was up. On attempting to restart the 'leader' node on the Consul cluster (as part of a Vault version upgrade) we encountered the following error (more detailed log fragments mentioned below):

==> Starting Consul agent...
==> Error starting agent: Failed to start Consul server: Failed to start Raft: recovery failed: failed to restore any of the available snapshots
bootstrap_expect > 0: expecting 3 servers

#### Reproduction Steps

When this issue happened on the 'leader' node, we attempted to restart the consul service on the other two nodes in the cluster as well and both of them failed to start with the exact same error message. Tried following the steps documented here - https://learn.hashicorp.com/consul/day-2-operations/outage (Failure of Multiple Servers in a Multi-Server Cluster), but that did not help either. 

We have an automated snapshot process which runs twice a week. So we tried provisioning a new Consul cluster (on a separate network) and then tried restoring the last snapshot and encountered this error:

Error restoring snapshot: Put https://10.1.0.11:8500/v1/snapshot: EOF

NOTE: The process which took this snapshot had no errors. It reported having successfully taken the snapshot, but restoration failed.

### Consul info for both Client and Server

<details>
  <summary>Client info</summary>

```
output from client 'consul info' command here
```

</details>

<details>
  <summary>Server info</summary>

```
agent:
	check_monitors = 0
	check_ttls = 1
	checks = 1
	services = 1
build:
	prerelease = 
	revision = 0bddfa23
	version = 1.4.0
consul:
	acl = disabled
	bootstrap = false
	known_datacenters = 1
	leader = true
	leader_addr = 10.1.0.11:8300
	server = true
raft:
	applied_index = 36311270
	commit_index = 36311270
	fsm_pending = 0
	last_contact = 0
	last_log_index = 36311270
	last_log_term = 4
	last_snapshot_index = 36303208
	last_snapshot_term = 4
	latest_configuration = [{Suffrage:Voter ID:5b067685-896a-4fda-9b25-aa3432e972db Address:10.1.0.11:8300} {Suffrage:Voter ID:3d30938f-a6bb-4338-9fd2-57e04db2c3ca Address:10.1.0.13:8300} {Suffrage:Voter ID:35f943ce-03a5-40b4-8fc2-1de3d7f870b8 Address:10.1.0.12:8300}]
	latest_configuration_index = 36268854
	num_peers = 2
	protocol_version = 3
	protocol_version_max = 3
	protocol_version_min = 0
	snapshot_version_max = 1
	snapshot_version_min = 0
	state = Leader
	term = 4
runtime:
	arch = amd64
	cpu_count = 2
	goroutines = 333
	max_procs = 2
	os = linux
	version = go1.11.1
serf_lan:
	coordinate_resets = 0
	encrypted = false
	event_queue = 0
	event_time = 570
	failed = 0
	health_score = 0
	intent_queue = 0
	left = 20
	member_time = 116268
	members = 95
	query_queue = 0
	query_time = 8278
serf_wan:
	coordinate_resets = 0
	encrypted = false
	event_queue = 0
	event_time = 1
	failed = 0
	health_score = 0
	intent_queue = 0
	left = 0
	member_time = 15
	members = 3
	query_queue = 0
	query_time = 1

```

</details>

### Operating system and Environment details

OS, Architecture, and any other information you can provide about the environment.

NAME="Ubuntu"
VERSION="16.04.6 LTS (Xenial Xerus)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 16.04.6 LTS"
VERSION_ID="16.04"
HOME_URL="http://www.ubuntu.com/"
SUPPORT_URL="http://help.ubuntu.com/"
BUG_REPORT_URL="http://bugs.launchpad.net/ubuntu/"
VERSION_CODENAME=xenial
UBUNTU_CODENAME=xenial

### Log Fragments

   2020/01/03 05:26:54 [INFO] agent: Deregistered service "vault-consul:10.1.0.11:8200"
    2020/01/03 05:26:54 [INFO] agent: Deregistered check "vault-consul:10.1.0.11:8200:vault-sealed-check"
    2020/01/03 05:26:55 [INFO] consul: server starting leave
    2020/01/03 05:26:55 [INFO] raft: Updating configuration with RemoveServer (491e5e2d-3970-466c-8acc-b2b4afc0cc4c, ) to [{Suffrage:Voter ID:b8feba76-93be-4c3e-b4cf-53a289090256 Address:10.1.0.12:8300} {Suffrage:Voter ID:d0ded429-e362-482f-a398-443cb30f4980 Address:10.1.0.13:8300}]
    2020/01/03 05:26:55 [INFO] raft: Removed ourself, transitioning to follower
    2020/01/03 05:26:55 [INFO] raft: Node at 10.1.0.11:8300 [Follower] entering Follower state (Leader: "")
    2020/01/03 05:26:55 [INFO] raft: aborting pipeline replication to peer {Nonvoter d0ded429-e362-482f-a398-443cb30f4980 10.1.0.13:8300}
    2020/01/03 05:26:55 [INFO] raft: aborting pipeline replication to peer {Voter b8feba76-93be-4c3e-b4cf-53a289090256 10.1.0.12:8300}
    2020/01/03 05:26:55 [INFO] consul: cluster leadership lost
    2020/01/03 05:26:56 [INFO] serf: EventMemberLeave: vault-consul-0.calculi 10.1.0.11
    2020/01/03 05:26:56 [INFO] consul: Handled member-leave event for server "vault-consul-0.calculi" in area "wan"
    2020/01/03 05:27:00 [INFO] serf: EventMemberLeave: vault-consul-0 10.1.0.11
    2020/01/03 05:27:00 [INFO] consul: Removing LAN server vault-consul-0 (Addr: tcp/10.1.0.11:8300) (DC: calculi)
    2020/01/03 05:27:04 [INFO] consul: Waiting 5s to drain RPC traffic
    2020/01/03 05:27:04 [ERR] agent: Coordinate update error: No cluster leader
    2020/01/03 05:27:04 [WARN] raft: not part of stable configuration, aborting election
    2020/01/03 05:27:09 [INFO] agent: Requesting shutdown
    2020/01/03 05:27:09 [INFO] consul: shutting down server
    2020/01/03 05:27:09 [INFO] manager: shutting down
    2020/01/03 05:27:09 [INFO] agent: consul server down
    2020/01/03 05:27:09 [INFO] agent: shutdown complete
    2020/01/03 05:27:09 [INFO] agent: Stopping DNS server 0.0.0.0:53 (tcp)
    2020/01/03 05:27:09 [INFO] agent: Stopping DNS server 0.0.0.0:53 (udp)
    2020/01/03 05:27:09 [INFO] agent: Stopping HTTPS server 0.0.0.0:8500 (tcp)
    2020/01/03 05:27:09 [INFO] agent: Waiting for endpoints to shut down
    2020/01/03 05:27:09 [INFO] agent: Endpoints down
    2020/01/03 05:27:09 [INFO] agent: Exit code: 0

==> Error starting agent: Failed to start Consul server: Failed to start Raft: recovery failed: failed to restore any of the available snapshots
bootstrap_expect > 0: expecting 3 servers
==> Starting Consul agent...
==> Error starting agent: Failed to start Consul server: Failed to start Raft: recovery failed: failed to restore any of the available snapshots
bootstrap_expect > 0: expecting 3 servers
==> Starting Consul agent...
==> Error starting agent: Failed to start Consul server: Failed to start Raft: recovery failed: failed to restore any of the available snapshots
bootstrap_expect > 0: expecting 3 servers
==> Starting Consul agent...
==> Error starting agent: Failed to start Consul server: Failed to start Raft: recovery failed: failed to restore any of the available snapshots

**Comment by pinaki124** [comment]
It seems the issue was caused by two nodes registering with the same name, one of them being 'Abc-0' and the other 'abc-0' (Notice the difference in case).

I got this error from the Consul server debug logs, when trying to restore a snapshot -

panic: failed to restore snapshot: failed to restore snapshot 2-36814322-1579194254301: check node "Abc-0" does not match node "abc-0"

goroutine 64 [running]:
github.com/hashicorp/consul/vendor/github.com/hashicorp/raft.(*Raft).restoreUserSnapshot(0xc0000ff080, 0xc0001e2070, 0x3809f20, 0xc00000c0f8, 0x0, 0x0)
	/go/src/github.com/hashicorp/consul/vendor/github.com/hashicorp/raft/raft.go:787 +0xd98
github.com/hashicorp/consul/vendor/github.com/hashicorp/raft.(*Raft).leaderLoop(0xc0000ff080)
	/go/src/github.com/hashicorp/consul/vendor/github.com/hashicorp/raft/raft.go:574 +0xab3
github.com/hashicorp/consul/vendor/github.com/hashicorp/raft.(*Raft).runLeader(0xc0000ff080)
	/go/src/github.com/hashicorp/consul/vendor/github.com/hashicorp/raft/raft.go:420 +0x34b
github.com/hashicorp/consul/vendor/github.com/hashicorp/raft.(*Raft).run(0xc0000ff080)
	/go/src/github.com/hashicorp/consul/vendor/github.com/hashicorp/raft/raft.go:140 +0x68
github.com/hashicorp/consul/vendor/github.com/hashicorp/raft.(*Raft).run-fm()
	/go/src/github.com/hashicorp/consul/vendor/github.com/hashicorp/raft/api.go:505 +0x2a
github.com/hashicorp/consul/vendor/github.com/hashicorp/raft.(*raftState).goFunc.func1(0xc0000ff080, 0xc000ba72c0)
	/go/src/github.com/hashicorp/consul/vendor/github.com/hashicorp/raft/state.go:146 +0x53
created by github.com/hashicorp/consul/vendor/github.com/hashicorp/raft.(*raftState).goFunc
	/go/src/github.com/hashicorp/consul/vendor/github.com/hashicorp/raft/state.go:144 +0x66

All Consul snapshots taken post the date when the new application, which was registering itself as 'Abc-0', came up, had the same issue.

**Comment by pinaki124** [comment]
Fixed by upgrading Consul to version 1.6.2
